File: Barchart.BinarySerializer/Types/BinarySerializerDateOnly.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) DateOnly values to (and from) a binary data source.
/// </summary>
public class BinarySerializerDateOnly : IBinaryTypeSerializer<DateOnly>
{
    #region Fields

    private readonly BinarySerializerInt _binarySerializerInt;
    
    #endregion

    #region Constructor(s)

    public BinarySerializerDateOnly()
    {
        _binarySerializerInt = new BinarySerializerInt();
    }
    
    #endregion

    #region Methods
    
    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, DateOnly value)
    {
        _binarySerializerInt.Encode(writer, GetDaysSinceEpoch(value));
    }

    /// <inheritdoc />
    public DateOnly Decode(IDataBufferReader reader)
    {
        int daysSinceEpoch = _binarySerializerInt.Decode(reader);
        
        return DateOnly.MinValue.AddDays(daysSinceEpoch);
    }
    
    /// <inheritdoc />
    public bool GetEquals(DateOnly a, DateOnly b)
    {
        return a.Equals(b);
    }
    
    private static int GetDaysSinceEpoch(DateOnly value)
    {
        return value.DayNumber - DateOnly.MinValue.DayNumber;
    }

    #endregion
}

File: typescript/src/types/binary-serializer-dateonly.ts
import { DataReader } from "../buffers/data-reader.interface";
import { DataWriter } from "../buffers/data-writer.interface";
import { BinarySerializerInt } from "./binary-serializer-int";
import { BinaryTypeSerializer } from "./binary-type-serializer.interface";
import Day from "@barchart/common-js/lang/Day";

/**
 * Reads (and writes) dateonly values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<Day>}
 */
export class BinarySerializerDateOnly implements BinaryTypeSerializer<Day> {
    private binarySerializerInt: BinarySerializerInt;

    constructor() {
        this.binarySerializerInt = new BinarySerializerInt();
    }

    get sizeInBytes(): number {
        return 4;
    }

    encode(writer: DataWriter, value: Day): void {
        this.binarySerializerInt.encode(writer, this.getDaysSinceEpoch(value));
    }

    decode(reader: DataReader): Day {
        const daysSinceEpoch = this.binarySerializerInt.decode(reader);
        return this.addDaysToEpoch(daysSinceEpoch);
    }

    getEquals(a: Day, b: Day): boolean {
        return a.getIsEqual(b);
    }

    getDaysSinceEpoch(value: Day): number {
        const epoch = new Day(1, 1, 1);
        return Day.countDaysBetween(epoch, value);
    }

    private addDaysToEpoch(days: number): Day {
        const epoch = new Day(1, 1, 1);
        return epoch.addDays(days);
    }
}